core/src/data/schema.rs

```diff
@@ -34,17 +35,128 @@ pub struct Schema {
     pub indexes: Vec<SchemaIndex>,
     pub engine: Option<String>,
     pub foreign_keys: Vec<ForeignKey>,
+    pub primary_key: Option<Vec<usize>>,
     pub comment: Option<String>,
 }
 
 impl Schema {
+    /// Returns the key associates to the provided row.
+    ///
+    /// # Arguments
+    /// * `row` - The row to get the key for.
+    pub fn get_primary_key<R: AsRef<[Value]>>(&self, row: R) -> Option<Key> {
+        self.primary_key.as_ref().map(|primary_key| {
+            let values = primary_key
+                .iter()
+                .map(|index| row.as_ref()[*index].clone())
+                .collect::<Vec<_>>();
+            if values.len() == 1 {
+                values.into_iter().next().unwrap().try_into().unwrap()
+            } else {
+                values.try_into().unwrap()
+            }
+        })
+    }
+
+    /// Returns an iterator over the ColumnDef instances in the schema that compose the primary key.
+    pub fn primary_key_columns(&self) -> Option<impl Iterator<Item = &ColumnDef>> {
+        self.primary_key.as_ref().map(|primary_key| {
+            primary_key.iter().filter_map(move |index| {
+                self.column_defs
+                    .as_ref()
+                    .and_then(|column_defs| column_defs.get(*index))
+            })
+        })
+    }
+
+    /// Returns an iterator over the ColumnDef instances in the schema that compose the primary key.
+    pub fn primary_key_column_names(&self) -> Option<impl Iterator<Item = &str>> {
+        self.primary_key_columns()
+            .map(|columns| columns.map(|column| column.name.as_str()))
+    }
+
+    /// Returns whether the schema has a primary key.
+    pub fn has_primary_key(&self) -> bool {
+        self.primary_key.is_some()
+    }
+
+    /// Returns the table name associated with the schema.
+    pub fn table_name(&self) -> &str {
+        &self.table_name
+    }
+
+    /// Returns whether the schema has column definitions.
+    pub fn has_column_defs(&self) -> bool {
+        self.column_defs.is_some()
+    }
```

Comment:
fields are public so could we just access by `self.primary_key`, `self.table_name` or `self.column_defs`?

Reply:
Will remove shortly

Reply:
Wait, I am not sure I understand - regarding table_name I agree, but the other two methods I would leave them be as they help with the code cleanliness and understandability.

Reply:
I have now removed `table_name`, but I would prefer leaving the other two methods.

Reply:
> Wait, I am not sure I understand - regarding table_name I agree, but the other two methods I would leave them be as they help with the code cleanliness and understandability.

if the `column_defs` and `primary_key` fields were private, then I am also for you. however, this is different.
I would think that this is against for code cleanliness, that we have two simple ways to achieve a single goal.
`has_column_defs()` or `self.column_defs.is_some()`.
someone may prefer `has_column_defs()` but others may not. and even someone thinks that it is better to use `self.column_defs.is_some()`.
without this kind of helper methods, we don't even need to care about this kind of selection issues.

if the helper method does a lot of roles so it is hard not to use that method, then that can be reasonable. but this case does not look like that.

Reply:
Ok, I have removed them.